File: backend/app/repositories/user_repository.py
"""User repository for database access"""
from typing import Optional
from app.models import db, User


class UserRepository:
    """Repository for User model database operations"""

    @staticmethod
    def create(email: str, name: str, password_hash: str) -> User:
        """Create a new user"""
        user = User(email=email, name=name)
        user.set_password(password_hash)
        db.session.add(user)
        db.session.commit()
        return user

    @staticmethod
    def create_google_user(email: str, name: str, google_id: str, profile_picture_url: Optional[str] = None) -> User:
        """Create a new user from Google OAuth"""
        user = User(
            email=email,
            name=name,
            oauth_provider='google',
            oauth_id=google_id,
            profile_picture=profile_picture_url
        )
        db.session.add(user)
        db.session.commit()
        return user

    @staticmethod
    def find_by_email(email: str) -> Optional[User]:
        """Find user by email"""
        return User.query.filter_by(email=email).first()

    @staticmethod
    def find_by_id(user_id: int) -> Optional[User]:
        """Find user by ID"""
        return User.query.get(user_id)

    @staticmethod
    def find_by_google_id(google_id: str) -> Optional[User]:
        """Find user by Google ID (oauth_id)"""
        return User.query.filter_by(oauth_provider='google', oauth_id=google_id).first()

    @staticmethod
    def update(user: User, **kwargs) -> User:
        """Update user fields"""
        # Map google_id to oauth_id for backward compatibility
        if 'google_id' in kwargs:
            kwargs['oauth_id'] = kwargs.pop('google_id')
            if 'oauth_provider' not in kwargs:
                kwargs['oauth_provider'] = 'google'
        if 'profile_picture_url' in kwargs:
            kwargs['profile_picture'] = kwargs.pop('profile_picture_url')
        
        for key, value in kwargs.items():
            if hasattr(user, key):
                setattr(user, key, value)
        db.session.commit()
        return user

    @staticmethod
    def deduct_tokens(user: User, tokens: float) -> User:
        """Deduct tokens from user account"""
        # Skip deduction for unlimited users
        if user.is_unlimited:
            return user
        if user.slide_tokens < tokens:
            raise ValueError('Insufficient tokens')
        user.slide_tokens -= tokens
        db.session.commit()
        return user

    @staticmethod
    def add_tokens(user: User, tokens: float) -> User:
        """Add tokens to user account"""
        user.slide_tokens += tokens
        db.session.commit()
        return user

File: backend/src/repositories/user.repository.ts
import bcrypt from 'bcryptjs';
import { and, eq, isNotNull } from 'drizzle-orm';
import { db } from '../db';
import { type User, accounts, users } from '../db/schema';

export class UserRepository {
  async findById(id: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.id, id));
    return user;
  }

  async findByEmail(email: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.email, email));
    return user;
  }

  async findByName(name: string): Promise<User | undefined> {
    const [user] = await db.select().from(users).where(eq(users.name, name));
    return user;
  }

  async deductTokens(userId: string, amount: number): Promise<void> {
    const user = await this.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    if (!user.isUnlimited && user.slideTokens < amount) {
      throw new Error('Insufficient tokens');
    }

    if (!user.isUnlimited) {
      await db
        .update(users)
        .set({
          slideTokens: user.slideTokens - amount,
        })
        .where(eq(users.id, userId));
    }
  }

  async awardDailyLoginBonus(userId: string): Promise<void> {
    const user = await this.findById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    const today = new Date().toISOString().split('T')[0];

    if (!user.lastLoginDate || user.lastLoginDate.toISOString().split('T')[0] !== today) {
      await db
        .update(users)
        .set({
          lastLoginDate: new Date(today),
          slideTokens: user.slideTokens + 10,
        })
        .where(eq(users.id, userId));
    }
  }

  async update(userId: string, data: Partial<User>): Promise<User | undefined> {
    const [updatedUser] = await db
      .update(users)
      .set({ ...data, updatedAt: new Date() })
      .where(eq(users.id, userId))
      .returning();
    return updatedUser;
  }

  async verifyPassword(user: User, password: string): Promise<boolean> {
    const [account] = await db
      .select()
      .from(accounts)
      .where(and(eq(accounts.userId, user.id), isNotNull(accounts.password)));

    if (!account || !account.password) return false;
    return await bcrypt.compare(password, account.password);
  }

  async updatePassword(userId: string, newPassword: string): Promise<void> {
    const hashedPassword = await bcrypt.hash(newPassword, 10);
    await db
      .update(accounts)
      .set({ password: hashedPassword })
      .where(and(eq(accounts.userId, userId), isNotNull(accounts.password)));
  }
}
